File: enterprise/backend/src/metabase_enterprise/advanced_permissions/api/impersonation.clj
(ns metabase-enterprise.advanced-permissions.api.impersonation
  (:require
   [compojure.core :refer [GET]]
   [metabase.api.common :as api]
   [metabase.util.malli.schema :as ms]
   [toucan2.core :as t2]))

(api/defendpoint GET "/"
  "Fetch a list of all Impersonation policies currently in effect, or a single policy if both `group_id` and `db_id`
  are provided."
  [group_id db_id]
  {group_id [:maybe ms/PositiveInt]
   db_id    [:maybe ms/PositiveInt]}
  (api/check-superuser)
  (if (and group_id db_id)
    (t2/select-one :model/ConnectionImpersonation :group_id group_id :db_id db_id)
    (t2/select :model/ConnectionImpersonation {:order-by [[:id :asc]]})))

(api/defendpoint DELETE "/:id"
  "Delete a Connection Impersonation entry."
  [id]
  {id ms/PositiveInt}
  (api/check-superuser)
  (api/check-404 (t2/select-one :model/ConnectionImpersonation :id id))
  (t2/delete! :model/ConnectionImpersonation :id id)
  api/generic-204-no-content)

(api/define-routes)

File: frontend/test/__support__/server-mocks/impersonation.ts
import fetchMock from "fetch-mock";

import type { DatabaseId, GroupId, Impersonation } from "metabase-types/api";

export const setupExistingImpersonationEndpoint = (
  impersonation: Impersonation,
) => {
  fetchMock.get(
    {
      url: `path:/api/ee/advanced-permissions/impersonation`,
      query: {
        db_id: impersonation.db_id,
        group_id: impersonation.group_id,
      },
    },
    impersonation,
  );
};

export const setupMissingImpersonationEndpoint = (
  databaseId: DatabaseId,
  groupId: GroupId,
) => {
  fetchMock.get(
    {
      url: `path:/api/ee/advanced-permissions/impersonation`,
      query: {
        db_id: databaseId,
        group_id: groupId,
      },
    },
    () => null,
  );
};
